// @rspress/core/dist/8995.421b464f.js
import { extname, join } from "./6289.29817fb4.js";
const PLUGIN_VIRTUAL_MODULE_NAME = 'rsbuild:virtual-module';
const pluginVirtualModule = (pluginOptions = {})=>({
        name: PLUGIN_VIRTUAL_MODULE_NAME,
        async setup (api) {
            const { virtualModules = {}, tempDir: virtualFolderName = '.rsbuild-virtual-module' } = pluginOptions;
            const TEMP_DIR = join(api.context.rootPath, virtualFolderName);
            const virtualFileAbsolutePaths = Object.keys(virtualModules).map((i)=>{
                let absolutePath = join(TEMP_DIR, i);
                if (!extname(absolutePath)) absolutePath = `${absolutePath}.js`;
                return [
                    i,
                    absolutePath
                ];
            });
            api.modifyBundlerChain((chain, { rspack })=>{
                const pluginId = 'RSBUILD_VIRTUAL_MODULE_PLUGIN';
                const virtualModules = Object.fromEntries(virtualFileAbsolutePaths.map((i)=>[
                        i[1],
                        ''
                    ]));
                if (chain.plugins.has(pluginId)) chain.plugin(pluginId).tap((options)=>{
                    const existingPaths = new Set(Object.keys(options[0] ?? {}));
                    const conflictPath = virtualFileAbsolutePaths.find(([, path])=>existingPaths.has(path));
                    if (conflictPath) throw new Error(`Virtual module name conflicted: ${conflictPath[0]}`);
                    return [
                        {
                            ...options[0],
                            ...virtualModules
                        }
                    ];
                });
                else chain.plugin(pluginId).use(rspack.experiments.VirtualModulesPlugin, [
                    virtualModules
                ]);
                chain.resolve.alias.merge(Object.fromEntries(virtualFileAbsolutePaths));
            });
            for (const [moduleName, absolutePath] of virtualFileAbsolutePaths){
                const handler = virtualModules[moduleName];
                if (handler) api.transform({
                    test: absolutePath
                }, handler);
            }
        }
    });
export { matchPath } from "react-router-dom";
export { DEFAULT_CONFIG_EXTENSIONS, DEFAULT_CONFIG_NAME, DEFAULT_PAGE_EXTENSIONS } from "@rspress/shared/constants";
export { glob } from "tinyglobby";
export { PLUGIN_REACT_NAME, pluginReact } from "@rsbuild/plugin-react";
export { chunk, groupBy } from "lodash-es";
export { default as remark_parse } from "remark-parse";
export { remove } from "unist-util-remove";
export { createReadStream, createWriteStream, default as node_fs, readFileSync } from "node:fs";
export { mergeRsbuildConfig, rspack as core_rspack } from "@rsbuild/core";
export { createHash } from "node:crypto";
export { pluginVirtualModule };
