// @rspress/core/dist/node/mdx/rehypePlugins/shiki.js
import { transformerAddLang, transformerAddLineNumbers, transformerAddTitle, createCssVariablesTheme, transformerAddWrapCode } from "../loader.15e0b2d2.js";
const cssVariablesTheme = createCssVariablesTheme({
    name: 'css-variables',
    variablePrefix: '--shiki-',
    variableDefaults: {},
    fontStyle: true
});
function createRehypeShikiOptions(showLineNumbers, defaultWrapCode, options) {
    const { transformers = [], ...restOptions } = options || {};
    const newTransformers = [
        transformerAddTitle(),
        transformerAddLang(),
        transformerAddLineNumbers({
            defaultShowLineNumbers: showLineNumbers
        }),
        transformerAddWrapCode({
            defaultWrapCode
        }),
        ...transformers
    ];
    return {
        theme: cssVariablesTheme,
        defaultLanguage: 'txt',
        lazy: true,
        langs: [
            'tsx',
            'ts',
            'js'
        ],
        ...restOptions,
        transformers: newTransformers
    };
}
export { createRehypeShikiOptions };
