// @rspress/core/dist/node/auto-nav-sidebar/utils.js
import { node_path, promises } from "../../6289.29817fb4.js";
import { loadFrontMatter, extractTextAndId } from "../../9348.1beb3e25.js";
async function pathExists(path) {
    try {
        await promises.access(path);
        return true;
    } catch  {
        return false;
    }
}
async function readJson(path) {
    const raw = await promises.readFile(path, 'utf8');
    return JSON.parse(raw);
}
async function extractInfoFromFrontmatterWithAbsolutePath(absolutePath, rootDir) {
    const fileHandle = await promises.open(absolutePath, 'r');
    try {
        const buffer = Buffer.alloc(2048);
        let content = '';
        let bytesRead = 0;
        let hasFrontmatter = false;
        let frontmatterEnd = -1;
        let h1Match = null;
        let frontmatter = {};
        do {
            const result = await fileHandle.read(buffer, 0, buffer.length, bytesRead);
            if (0 === result.bytesRead) break;
            content += buffer.subarray(0, result.bytesRead).toString('utf-8');
            bytesRead += result.bytesRead;
            if (bytesRead <= buffer.length) if (content.startsWith('---')) {
                hasFrontmatter = true;
                const secondDelimiter = content.indexOf('\n---\n', 4);
                if (-1 !== secondDelimiter) {
                    frontmatterEnd = secondDelimiter + 5;
                    const { frontmatter: fm } = loadFrontMatter(content.slice(0, frontmatterEnd + 100), absolutePath, rootDir);
                    frontmatter = fm;
                }
            } else hasFrontmatter = false;
            if (!h1Match) {
                const searchStart = hasFrontmatter && frontmatterEnd > -1 ? frontmatterEnd : 0;
                const h1RegExp = /^#\s+(.*?)\n$/m;
                h1Match = content.slice(searchStart).match(h1RegExp);
            }
            if (hasFrontmatter && frontmatterEnd > -1 && h1Match) break;
            if (!hasFrontmatter && h1Match) break;
            if (hasFrontmatter && frontmatterEnd > -1 && bytesRead > 16384) break;
        }while (bytesRead < 65536);
        if (hasFrontmatter && -1 === frontmatterEnd) {
            const { frontmatter: fm } = loadFrontMatter(content, absolutePath, rootDir);
            frontmatter = fm;
        }
        if (!h1Match) {
            const h1RegExp = /^#\s+(.*)$/m;
            h1Match = content.match(h1RegExp);
        }
        const fileNameWithoutExt = node_path.basename(absolutePath, node_path.extname(absolutePath));
        return {
            title: extractTextAndId(frontmatter.title || h1Match?.[1] || fileNameWithoutExt)[0],
            overviewHeaders: frontmatter.overviewHeaders,
            context: frontmatter.context,
            tag: frontmatter.tag,
            description: frontmatter.description
        };
    } finally{
        await fileHandle.close();
    }
}
export { extractInfoFromFrontmatterWithAbsolutePath, pathExists, readJson };
