// @rspress/core/dist/_private/react/reconciler.js
import { createContext } from "react";
import react_reconciler from "react-reconciler";
import { DefaultEventPriority, NoEventPriority } from "react-reconciler/constants.js";
class TextNode {
    text;
    parent;
    constructor(text){
        this.text = text;
    }
    setText(text) {
        this.text = text;
    }
}
class MarkdownNode {
    type;
    props;
    children;
    parent;
    constructor(type, props = {}){
        this.type = type;
        this.props = props;
        this.children = [];
    }
    appendChild(child) {
        this.children.push(child);
        if ('parent' in child) child.parent = this;
    }
    removeChild(child) {
        const index = this.children.indexOf(child);
        if (-1 !== index) this.children.splice(index, 1);
    }
    insertBefore(child, beforeChild) {
        const index = this.children.indexOf(beforeChild);
        if (-1 !== index) this.children.splice(index, 0, child);
        else this.appendChild(child);
        if ('parent' in child) child.parent = this;
    }
}
let currentUpdatePriority = NoEventPriority;
const hostConfig = {
    supportsMutation: true,
    supportsPersistence: false,
    supportsHydration: false,
    isPrimaryRenderer: false,
    getRootHostContext () {
        return {
            isInsideText: false
        };
    },
    getChildHostContext (parentHostContext, type) {
        const previousIsInsideText = parentHostContext.isInsideText;
        const isInsideText = 'text' === type || 'span' === type || previousIsInsideText;
        if (previousIsInsideText === isInsideText) return parentHostContext;
        return {
            isInsideText
        };
    },
    createInstance (type, props) {
        return new MarkdownNode(type, props);
    },
    createTextInstance (text) {
        return new TextNode(text);
    },
    shouldSetTextContent (_type, _props) {
        return false;
    },
    appendChild (parent, child) {
        parent.appendChild(child);
    },
    insertBefore (parent, child, beforeChild) {
        parent.insertBefore(child, beforeChild);
    },
    removeChild (parent, child) {
        parent.removeChild(child);
    },
    commitTextUpdate (textInstance, _oldText, newText) {
        textInstance.setText(newText);
    },
    commitUpdate (instance, _type, _oldProps, newProps, _internalHandle) {
        instance.props = newProps;
    },
    finalizeInitialChildren () {
        return false;
    },
    commitMount () {},
    clearContainer (container) {
        container.children = [];
    },
    setCurrentUpdatePriority (newPriority) {
        currentUpdatePriority = newPriority;
    },
    getCurrentUpdatePriority () {
        return currentUpdatePriority;
    },
    resolveUpdatePriority () {
        if (currentUpdatePriority !== NoEventPriority) return currentUpdatePriority;
        return DefaultEventPriority;
    },
    appendInitialChild (parent, child) {
        parent.appendChild(child);
    },
    preparePortalMount () {},
    resetTextContent () {},
    getPublicInstance (instance) {
        return instance;
    },
    prepareForCommit (_containerInfo) {
        return null;
    },
    resetAfterCommit (_containerInfo) {},
    scheduleTimeout: setTimeout,
    cancelTimeout: clearTimeout,
    noTimeout: -1,
    scheduleMicrotask: 'function' == typeof queueMicrotask ? queueMicrotask : (fn)=>Promise.resolve().then(fn),
    getInstanceFromNode () {
        return null;
    },
    beforeActiveInstanceBlur () {},
    afterActiveInstanceBlur () {},
    prepareScopeUpdate () {},
    getInstanceFromScope () {
        return null;
    },
    detachDeletedInstance () {},
    appendChildToContainer (container, child) {
        container.appendChild(child);
    },
    insertInContainerBefore (container, child, beforeChild) {
        container.insertBefore(child, beforeChild);
    },
    removeChildFromContainer (container, child) {
        container.removeChild(child);
    },
    hideInstance (_instance) {},
    hideTextInstance (textInstance) {
        textInstance.setText('');
    },
    unhideInstance (_instance) {},
    unhideTextInstance (textInstance, text) {
        textInstance.setText(text);
    },
    maySuspendCommit () {
        return false;
    },
    preloadInstance () {
        return true;
    },
    startSuspendingCommit () {},
    suspendInstance () {},
    waitForCommitToBeReady () {
        return null;
    },
    NotPendingTransition: null,
    HostTransitionContext: createContext(null),
    resetFormInstance () {},
    requestPostPaintCallback () {},
    shouldAttemptEagerTransition () {
        return false;
    },
    trackSchedulerEvent () {},
    resolveEventType () {
        return null;
    },
    resolveEventTimeStamp () {
        return -1.1;
    }
};
const reconciler = react_reconciler(hostConfig);
export { MarkdownNode, TextNode, reconciler };
