// @rspress/core/dist/node/utils/reactAlias.js
import { node_path, logger, picocolors } from "../../6289.29817fb4.js";
import { core_rspack } from "../../8995.421b464f.js";
import { pathExists, readJson } from "./fs.js";
import { PACKAGE_ROOT } from "../constants.js";
import { hintReactVersion } from "../logger/hint.js";
const Resolver = core_rspack.experiments.resolver.ResolverFactory;
async function detectPackageMajorVersion(name) {
    const cwd = process.cwd();
    const pkgPath = node_path.join(cwd, 'node_modules', name);
    if (await pathExists(pkgPath)) {
        const pkgJson = await readJson(node_path.join(pkgPath, 'package.json'));
        const version = Number(pkgJson.version?.split('.')[0]);
        return version;
    }
}
async function resolveReactRouterDomAlias() {
    const hasInstalled = await detectPackageMajorVersion('react-router-dom');
    const basedir = hasInstalled ? process.cwd() : PACKAGE_ROOT;
    const alias = {};
    const resolver = new Resolver({
        mainFields: [
            'browser',
            'module',
            'main'
        ],
        extensions: [
            '.js'
        ],
        alias
    });
    try {
        const resolved = await resolver.async(basedir, 'react-router-dom/package.json');
        if (resolved.error) throw Error(resolved.error);
        if (!resolved.path) throw Error("'react-router-dom' resolved to empty path");
        return {
            'react-router-dom': node_path.dirname(resolved.path)
        };
    } catch (e) {
        logger.warn('react-router-dom not found: \n', e);
    }
    return {};
}
async function resolveReactAlias(isSSR) {
    const hasInstalled = await detectPackageMajorVersion('react');
    const basedir = hasInstalled ? process.cwd() : PACKAGE_ROOT;
    const libPaths = [
        'react',
        'react/jsx-runtime',
        'react/jsx-dev-runtime',
        'react/package.json',
        'react-dom',
        'react-dom/client',
        'react-dom/server',
        'react-dom/package.json'
    ];
    const alias = {};
    const resolver = new Resolver({
        extensions: [
            '.js'
        ],
        alias,
        conditionNames: isSSR ? [
            '...'
        ] : [
            'browser',
            '...'
        ]
    });
    await Promise.all(libPaths.map(async (lib)=>{
        try {
            const resolved = await resolver.async(basedir, lib);
            if (resolved.error || !resolved.path) throw Error(resolved.error);
            alias[lib] = resolved.path;
        } catch (e) {
            if (e instanceof Error) {
                logger.warn(`${lib} not found: \n    ${picocolors.gray(e.toString())}`);
                hintReactVersion();
            }
        }
    }));
    return alias;
}
export { resolveReactAlias, resolveReactRouterDomAlias };
