// @rspress/core/dist/node/mdx/remarkPlugins/image.js
import { isExternalUrl, getNodeAttribute, visit } from "../../../9348.1beb3e25.js";
import { getDefaultImportAstNode } from "../../utils/getImportAstNode.js";
const normalizeImageUrl = (imageUrl)=>{
    if (isExternalUrl(imageUrl) || imageUrl.startsWith('/')) return '';
    return imageUrl;
};
const remarkImage = ()=>(tree, _file)=>{
        const images = [];
        const getMdxSrcAttribute = (tempVar)=>({
                type: 'mdxJsxAttribute',
                name: 'src',
                value: {
                    type: 'mdxJsxAttributeValueExpression',
                    value: tempVar,
                    data: {
                        estree: {
                            type: 'Program',
                            sourceType: 'module',
                            body: [
                                {
                                    type: 'ExpressionStatement',
                                    expression: {
                                        type: 'Identifier',
                                        name: tempVar
                                    }
                                }
                            ]
                        }
                    }
                }
            });
        visit(tree, 'image', (node)=>{
            const { alt, url } = node;
            if (!url) return;
            const imagePath = normalizeImageUrl(url);
            if (!imagePath) return;
            const tempVariableName = `image${images.length}`;
            Object.assign(node, {
                type: 'mdxJsxFlowElement',
                name: 'img',
                children: [],
                attributes: [
                    alt && {
                        type: 'mdxJsxAttribute',
                        name: 'alt',
                        value: alt
                    },
                    getMdxSrcAttribute(tempVariableName)
                ].filter(Boolean)
            });
            images.push(getDefaultImportAstNode(tempVariableName, imagePath));
        });
        visit(tree, (node)=>{
            if ('mdxJsxFlowElement' !== node.type && 'mdxJsxTextElement' !== node.type || 'img' !== node.name) return;
            const srcAttr = getNodeAttribute(node, 'src', true);
            if ('string' != typeof srcAttr?.value) return;
            const imagePath = normalizeImageUrl(srcAttr.value);
            if (!imagePath) return;
            const tempVariableName = `image${images.length}`;
            Object.assign(srcAttr, getMdxSrcAttribute(tempVariableName));
            images.push(getDefaultImportAstNode(tempVariableName, imagePath));
        });
        tree.children.unshift(...images);
    };
export { remarkImage };
