// @rspress/core/dist/node/ssg/renderPages.js
/*! For license information please see renderPages.js.LICENSE.txt */
import { logger, picocolors } from "../../6289.29817fb4.js";
import { chunk } from "../../8995.421b464f.js";
import { SSGWorkerThreadTaskSize, getNumberOfThreads, SSGWorkerThreadRecyclerMaxMemory, SSGConcurrency } from "./ssgEnv.js";
import { pMap } from "../../5485.888e77db.js";
import { renderPage } from "./renderPage.js";
import { hintSSGFailed } from "../logger/hint.js";
import { renderHtmlTemplate } from "./renderHtmlTemplate.js";
const routePath2HtmlFileName = (routePath)=>{
    let fileName = routePath;
    fileName = fileName.endsWith('/') ? `${routePath}index.html` : `${routePath}.html`;
    return fileName.replace(/^\/+/, '');
};
async function renderPages(routeService, config, ssrBundlePath, htmlTemplate, emitAsset) {
    logger.info('Rendering pages...');
    const startTime = Date.now();
    const ssg = config.ssg || true;
    try {
        const routes = routeService.getRoutes();
        if (!routeService.isExistRoute('/404')) routes.push({
            routePath: '/404'
        });
        if ('object' == typeof ssg && Array.isArray(ssg?.experimentalExcludeRoutePaths) && ssg.experimentalExcludeRoutePaths.length > 0) {
            const shouldIgnoreRoutePaths = new Set();
            const shouldIgnoreRoutePathsRegExp = [];
            for (const routePath of ssg.experimentalExcludeRoutePaths)if ('string' == typeof routePath) shouldIgnoreRoutePaths.add(routePath);
            else if (routePath instanceof RegExp) shouldIgnoreRoutePathsRegExp.push(routePath);
            const promiseList = [];
            const csrRoutes = [];
            const ssgRoutes = [];
            for (const route of routes)if (shouldIgnoreRoutePaths.has(route.routePath) || shouldIgnoreRoutePathsRegExp.some((reg)=>reg.test(route.routePath))) {
                csrRoutes.push(route);
                promiseList.push(renderCSRPage(config.head, route, htmlTemplate, emitAsset));
            } else ssgRoutes.push(route);
            if (csrRoutes.length > 0) {
                const csrRoutePaths = csrRoutes.map((r)=>r.routePath);
                const MAX_DISPLAY_ROUTES = 10;
                const displayedRoutes = csrRoutePaths.slice(0, MAX_DISPLAY_ROUTES);
                const remainingCount = csrRoutePaths.length - MAX_DISPLAY_ROUTES;
                logger.info(`Some routes (${picocolors.yellow(csrRoutePaths.length)} total) are ignored in SSG and fallback to CSR via \`ssg.experimentalExcludeRoutePaths\`: ${picocolors.yellow(displayedRoutes.join(', '))}${remainingCount > 0 ? picocolors.dim(` ... and ${remainingCount} more`) : ''}`);
                await Promise.all(promiseList);
                routes.length = 0;
                routes.push(...ssgRoutes);
            }
        }
        if ('object' == typeof ssg && ssg?.experimentalWorker) {
            const numberOfThreads = getNumberOfThreads(routes.length);
            const Tinypool = await import("tinypool").then((m)=>m.default);
            const pool = new Tinypool({
                filename: new URL('./renderPageWorker.js', import.meta.url).href,
                concurrentTasksPerWorker: 1,
                minThreads: numberOfThreads,
                maxThreads: numberOfThreads,
                /**
         * https://github.com/facebook/docusaurus/blob/0306d182407bb98f140cd5ec7481fa9608fe0297/packages/docusaurus/src/ssg/ssgExecutor.ts#L123
         * @license MIT
         */ maxMemoryLimitBeforeRecycle: SSGWorkerThreadRecyclerMaxMemory(),
                isolateWorkers: false,
                workerData: {
                    params: {
                        htmlTemplate,
                        head: config.head,
                        ssrBundlePath
                    }
                }
            });
            await Promise.all(chunk(routes, SSGWorkerThreadTaskSize()).map(async (routes)=>{
                const htmlList = await pool.run({
                    routes
                });
                for(let i = 0; i < routes.length; i++){
                    const route = routes[i];
                    const html = htmlList[i];
                    const fileName = routePath2HtmlFileName(route.routePath);
                    emitAsset(fileName, html);
                }
            }));
            await pool.destroy();
        } else await pMap(routes, async (route)=>{
            const html = await renderPage(route, htmlTemplate, config.head, ssrBundlePath);
            const fileName = routePath2HtmlFileName(route.routePath);
            emitAsset(fileName, html);
        }, {
            concurrency: SSGConcurrency()
        });
        const totalTime = Date.now() - startTime;
        logger.success(`Pages rendered in ${picocolors.yellow(totalTime)} ms.`);
    } catch (e) {
        if (e instanceof Error) {
            logger.error(`Pages render error: ${e.message}`);
            logger.debug(e);
            hintSSGFailed();
        }
        throw e;
    }
}
async function renderCSRPage(head, route, htmlTemplate, emitAsset) {
    const html = await renderHtmlTemplate(htmlTemplate, head, route, '');
    const fileName = routePath2HtmlFileName(route.routePath);
    emitAsset(fileName, html);
}
async function renderCSRPages(routeService, config, htmlTemplate, emitAsset) {
    const routes = routeService.getRoutes();
    if (!routeService.isExistRoute('/404')) routes.push({
        routePath: '/404'
    });
    await Promise.all(routes.map((route)=>renderCSRPage(config.head, route, htmlTemplate, emitAsset)));
}
export { renderCSRPages, renderPages };
