// @rspress/core/dist/runtime/index.js
import { isDataUrl, isExternalUrl, matchNavbar, matchSidebar, normalizeHref } from "@rspress/shared";
import { Head } from "@unhead/react";
import { Content } from "./Content.js";
import { useActiveMatcher } from "./hooks/useActiveMatcher.js";
import { ThemeContext, useDark } from "./hooks/useDark.js";
import { useFrontmatter } from "./hooks/useFrontmatter.js";
import { useI18n } from "./hooks/useI18n.js";
import { useLang } from "./hooks/useLang.js";
import { useLocaleSiteData } from "./hooks/useLocaleSiteData.js";
import { useNav } from "./hooks/useNav.js";
import { PageContext, usePage } from "./hooks/usePage.js";
import { usePageData } from "./hooks/usePageData.js";
import { usePages } from "./hooks/usePages.js";
import { getSidebarDataGroup, useSidebar, useSidebarDynamic } from "./hooks/useSidebar.js";
import { useSite } from "./hooks/useSite.js";
import { useVersion } from "./hooks/useVersion.js";
import { useWindowSize } from "./hooks/useWindowSize.js";
import { NoSSR } from "./NoSSR.js";
import { isActive, pathnameToRouteService, preloadLink } from "./route.js";
import { addLeadingSlash, addTrailingSlash, cleanUrlByConfig, isEqualPath, isProduction, normalizeHrefInRuntime, normalizeImagePath, removeBase, removeTrailingSlash, routePathToMdPath, withBase } from "./utils.js";
export * from "react-router-dom";
export { Content, Head, NoSSR, PageContext, ThemeContext, addLeadingSlash, addTrailingSlash, cleanUrlByConfig, getSidebarDataGroup, isActive, isDataUrl, isEqualPath, isExternalUrl, isProduction, matchNavbar, matchSidebar, normalizeHref, normalizeHrefInRuntime, normalizeImagePath, pathnameToRouteService, preloadLink, removeBase, removeTrailingSlash, routePathToMdPath, useActiveMatcher, useDark, useFrontmatter, useI18n, useLang, useLocaleSiteData, useNav, usePage, usePageData, usePages, useSidebar, useSidebarDynamic, useSite, useVersion, useWindowSize, withBase };
