// @rspress/core/dist/node/runtimeModule/siteData/createSiteData.js
import { getIconUrlPath } from "../../../9348.1beb3e25.js";
import { normalizeThemeConfig } from "./normalizeThemeConfig.js";
async function createSiteData(userConfig) {
    const tempSearchObj = Object.assign({}, userConfig.search);
    if (tempSearchObj) tempSearchObj.searchHooks = void 0;
    const siteData = {
        base: userConfig.base ?? '/',
        title: userConfig?.title || '',
        description: userConfig?.description || '',
        icon: getIconUrlPath(userConfig?.icon) || '',
        route: userConfig?.route || {},
        themeConfig: await normalizeThemeConfig(userConfig),
        lang: userConfig?.lang || '',
        locales: userConfig?.locales || userConfig.themeConfig?.locales || [],
        logo: userConfig?.logo || '',
        logoText: userConfig?.logoText || '',
        logoHref: userConfig?.logoHref || '',
        multiVersion: {
            default: userConfig?.multiVersion?.default || '',
            versions: userConfig?.multiVersion?.versions || []
        },
        search: tempSearchObj ?? {
            mode: 'local'
        },
        markdown: {
            showLineNumbers: userConfig?.markdown?.showLineNumbers ?? false,
            defaultWrapCode: userConfig?.markdown?.defaultWrapCode ?? false,
            shiki: {}
        }
    };
    return {
        siteData
    };
}
export { createSiteData };
