// @rspress/core/dist/node/ssg-md/rsbuildPluginSSGMD.js
import { writeFile, mkdir, rm, join } from "../../6289.29817fb4.js";
import { isDebugMode } from "../../9348.1beb3e25.js";
import { NODE_SSG_MD_BUNDLE_FOLDER, NODE_SSG_MD_BUNDLE_NAME } from "../constants.js";
import { renderPages } from "./renderPages.js";
import { emitLlmsTxt } from "./llms/emitLlmsTxt.js";
const rsbuildPluginSSGMD = ({ routeService, config })=>({
        name: 'rspress-inner-rsbuild-plugin-ssg',
        async setup (api) {
            api.onBeforeBuild(()=>{
                let hasError = false;
                api.processAssets({
                    stage: 'optimize-transfer',
                    environments: [
                        'node_md'
                    ]
                }, async ({ assets, compilation, environment, compiler })=>{
                    if (compilation.errors.length > 0) {
                        hasError = true;
                        return;
                    }
                    if (hasError) return;
                    const emitAsset = (assetName, content)=>{
                        compilation.emitAsset(assetName, new compiler.webpack.sources.RawSource(content));
                    };
                    const distPath = environment.distPath;
                    const ssgFolderPath = join(distPath, NODE_SSG_MD_BUNDLE_FOLDER);
                    const mainCjsAbsolutePath = join(ssgFolderPath, NODE_SSG_MD_BUNDLE_NAME);
                    await mkdir(ssgFolderPath, {
                        recursive: true
                    });
                    await Promise.all(Object.entries(assets).map(async ([assetName, assetSource])=>{
                        if (assetName.startsWith(`${NODE_SSG_MD_BUNDLE_FOLDER}/`)) {
                            const fileAbsolutePath = join(distPath, assetName);
                            await writeFile(fileAbsolutePath, assetSource.source().toString());
                            compilation.deleteAsset(assetName);
                        }
                    }));
                    const mdContents = await renderPages(routeService, mainCjsAbsolutePath, emitAsset);
                    await emitLlmsTxt(config, routeService, emitAsset, mdContents);
                    if (!isDebugMode()) await rm(ssgFolderPath, {
                        recursive: true
                    });
                });
            });
        }
    });
export { rsbuildPluginSSGMD };
