// @rspress/core/dist/theme/components/CodeBlock/index.js
import { Fragment, jsx, jsxs } from "react/jsx-runtime";
import { CodeButtonGroup, useCodeButtonGroup } from "@theme";
import clsx from "clsx";
import "./index.css";
function CodeBlock({ containerElementClassName, title, lang = 'txt', wrapCode: wrapCodeProp = false, lineNumbers: lineNumbersProp = false, codeButtonGroupProps, children }) {
    if (process.env.__SSR_MD__) return /*#__PURE__*/ jsx(Fragment, {
        children: children
    });
    const { codeWrap, toggleCodeWrap, copyElementRef } = useCodeButtonGroup(wrapCodeProp);
    return /*#__PURE__*/ jsxs("div", {
        className: clsx('rp-codeblock', `language-${lang}`, containerElementClassName),
        children: [
            title && /*#__PURE__*/ jsx("div", {
                className: "rp-codeblock__title",
                children: title
            }),
            /*#__PURE__*/ jsxs("div", {
                className: clsx('rp-codeblock__content', codeWrap && 'rp-codeblock__content--code-wrap', lineNumbersProp && 'rp-codeblock__content--line-numbers'),
                children: [
                    /*#__PURE__*/ jsx("div", {
                        className: "rp-codeblock__content__scroll-container rp-scrollbar rp-scrollbar--always",
                        ref: copyElementRef,
                        children: children
                    }),
                    /*#__PURE__*/ jsx(CodeButtonGroup, {
                        ...codeButtonGroupProps,
                        copyElementRef: copyElementRef,
                        codeWrap: codeWrap,
                        toggleCodeWrap: toggleCodeWrap
                    })
                ]
            })
        ]
    });
}
export { CodeBlock };
